### ISSUE TITLE
bug: Converted-sloppak stems sound "tinny": double-lossy encode + htdemucs_6s artifacts

### ISSUE THREAD

**Issue by mogul** [got-feedBack/feedBack#470]
> _Restored from `slopsmith/slopsmith#340` — original issue, opened by @topkoa on 2026-05-20._
> _[restored-from: slopsmith/slopsmith#340]_

## Summary

Multiple users report that [redacted] → sloppak converted songs sound "tinny" when played back through Slopsmith. One user verified that running `demucs -n htdemucs_ft "song.mp3"` manually produces stems that are "almost indistinguishable from the original" when summed, which suggests the regression is in our conversion pipeline (model choice + encoding), not in demucs itself.

This issue documents the full pipeline, the most likely root causes, and a ranked list of fixes. No code changes proposed yet — opening for discussion before picking a direction.

## The conversion pipeline (verified)

[redacted] → sloppak with stems:

1. **[redacted] WEM extraction** — [lib/sloppak_convert.py#L346-L350](../blob/main/lib/sloppak_convert.py#L346-L350): `_wem_to_ogg(wems[0], work_dir/"stems"/"full.ogg")`.
2. **WEM → WAV (lossless)** — [lib/sloppak_convert.py#L84](../blob/main/lib/sloppak_convert.py#L84): vgmstream-cli decodes WEM to a temp WAV. No loss here.
3. **WAV → OGG (Lossy #338)** — [lib/audio.py#L268-L269](../blob/main/lib/audio.py#L268-L269):
   \`\`\`python
   [ffmpeg, "-y", "-i", str(wav), "-c:a", "libvorbis", "-q:a", "5", str(out_ogg)]
   \`\`\`
   The lossless WAV is re-encoded to Vorbis q:a 5 (~160 kbps nominal) as \`stems/full.ogg\`.
4. **Demucs split** — [lib/sloppak_convert.py#L580-L581](../blob/main/lib/sloppak_convert.py#L580-L581):
   \`\`\`python
   cmd = [sys.executable, "-c", bootstrap, json.dumps(extra_paths),
          "-n", model, "-o", str(out_dir), str(full_ogg)]
   \`\`\`
   **Demucs is fed the Vorbis-compressed \`full.ogg\`, not the lossless WAV from step 2.** The model defaults to \`htdemucs_6s\` ([line 689](../blob/main/lib/sloppak_convert.py#L689)). No \`--shifts\`, \`--overlap\`, or \`--segment\` overrides — defaults only (shifts=1).
5. **Stem WAV → OGG (Lossy #2)** — [lib/sloppak_convert.py#L670](../blob/main/lib/sloppak_convert.py#L670) → [lib/audio.py#L268](../blob/main/lib/audio.py#L268): each demucs WAV stem is re-encoded at Vorbis q:a 5.
6. **\`full.ogg\` deleted** — [lib/sloppak_convert.py#L683](../blob/main/lib/sloppak_convert.py#L683). The final sloppak ships only the 6 split stems at q:a 5; the browser sums them at playback time.

User's manual pipeline (sounds great): \`mp3 → demucs htdemucs_ft → WAV stems → play uncompressed\`.

### 1. \`htdemucs_6s\` model artifacts

The reporting user explicitly compared against \`htdemucs_ft\` and found it indistinguishable from the original. Slopsmith uses \`htdemucs_6s\`:

- \`htdemucs_ft\` — 4-source (vocals/drums/bass/other), fine-tuned, generally considered the highest-quality demucs model. No guitar/piano split.
- \`htdemucs_6s\` — 6-source (adds guitar + piano), trained on a smaller specialized dataset, and known to produce more separation artifacts than \`_ft\`: residual hiss, smeared transients, treble bleed. The choice is intentional and documented in [plugins/sloppak_converter/README.md](../blob/main/plugins/sloppak_converter/README.md) because guitar-focused users want the guitar/piano stems.

The "tinny" complaint matches the known artifact profile of \`_6s\`: HF separation residuals that sound like cymbals + amp hiss leaking across stems.

### 2. Vorbis q:a 5 applied per-stem compounds when summed

\`-q:a 5\` is ~160 kbps. On a single full mix this is mild compression most listeners can't identify blind. But:

- The pipeline encodes **6 stems independently** at q:a 5, then the browser sums them at playback.
- Vorbis quantization noise is decorrelated between stems → noise sums incoherently → effective SNR is worse than a single q:a 5 mix of the same material.
- Demucs stems already carry HF separation artifacts; q:a 5 then under-allocates bits to HF in each stem → cumulative HF degradation → "tinny."

### 3. Demucs sees a Vorbis-compressed input

\`full.ogg\` is q:a 5 Vorbis. Demucs decodes it back to float32 and separates on already-degraded HF content. Separation quality drops when the input is lossy (the demucs community generally recommends FLAC/WAV input).

A trivial fix: pass the lossless WAV from step 2 directly to demucs, then encode stems. Saves time too — no WEM→OGG roundtrip.

### 4. libvorbis fallback silently degrades

[lib/audio.py#L274-L280](../blob/main/lib/audio.py#L274-L280): if the ffmpeg build lacks \`libvorbis\`, the code falls back to ffmpeg's built-in \`vorbis\` encoder under \`-strict experimental\`. The inline comment explicitly notes "produces a lower-quality file." No log, no warning surfaced to the user. Worth checking how many real-world installs hit this path.

### 5. Demucs \`--shifts\` not exposed

Default \`--shifts=1\` = single forward pass. \`--shifts=2\` averages two shifted passes (cleaner separation, ~2× CPU cost). Standard quality knob.

## Quick reproduction / verification for users

- \`ffprobe stems/guitar.ogg\` on a converted sloppak — confirms ~160 kbps Vorbis and the sample rate.
- Open a stem in Audacity → Spectrogram view → look for the hard rolloff above ~16 kHz (Vorbis q:a 5 signature) AND the blocky HF artifacts characteristic of htdemucs_6s.
- Run \`demucs -n htdemucs_ft <full.ogg>\` manually on the same ogg Slopsmith fed in, then compare against Slopsmith's stems. This isolates the model contribution from the encoding contribution.

## Proposed fixes (ranked)

| # | Change | Effort | Impact | Trade-off |
|---|--------|--------|--------|-----------|
| A | Skip Lossy #338 — feed lossless WAV to demucs, delete after split | Tiny | Medium | None (pure win, also faster) |
| B | Bump stem Vorbis to \`-q:a 7\` (~224 kbps) or \`-q:a 8\` (~256) | Tiny | Medium-high | Stems ~40-60% larger on disk |
| C | Surface a libvorbis-missing warning at convert time | Small | Diagnostic | None |
| D | Default to \`--shifts 2\` | Tiny code, big CPU | Medium | 2× demucs runtime |
| E | Offer \`htdemucs_ft\` as a model option in the settings UI (loses guitar/piano split — falls into \`other\`) | Medium | High when picked | UI work + schema field for model preference |
| F | Switch stem encoding to FLAC (lossless) — eliminates Lossy #2 | Small | Highest audio impact | Files ~3-5× larger; browser FLAC playback is fine |

\`A + B\` together would address most of the perceived tinniness without changing the model. \`E\` is what the original reporter actually wants but breaks the guitar-stem feature.

## Critical files

- [lib/sloppak_convert.py#L74-L94](../blob/main/lib/sloppak_convert.py#L74-L94) — \`_wem_to_ogg\` (Lossy #338)
- [lib/sloppak_convert.py#L580-L581](../blob/main/lib/sloppak_convert.py#L580-L581) — demucs invocation
- [lib/sloppak_convert.py#L603-L611](../blob/main/lib/sloppak_convert.py#L603-L611) — \`_encode_ogg\` (Lossy #2 entry)
- [lib/sloppak_convert.py#L689](../blob/main/lib/sloppak_convert.py#L689) — default model
- [lib/audio.py#L262-L280](../blob/main/lib/audio.py#L262-L280) — \`_ffmpeg_wav_to_ogg\` (single q:a 5 site, shared by both lossy steps)
- [plugins/sloppak_converter/routes.py#L115-L120](../blob/main/plugins/sloppak_converter/routes.py#L115-L120) — plugin entry point (where a model setting would land)

## Verification plan (when a fix lands)

1. Convert the in-tree note_detect benchmark sloppak with old vs. new pipeline.
2. \`ffprobe\` each stem — confirm new bitrate / codec.
3. Spectrogram-compare a guitar stem before/after.
4. A/B blind listen against the original WEM-extracted mix.
5. Time the conversion — \`--shifts 2\` and lossless intermediates change runtime materially.

**Comment by mogul** [comment]
_@barlind commented on 2026-05-21:_

Perhaps we should add additional stems metadata in the sloppak for the model and engine used?
